server/tests/src/main/scala/sttp/tapir/server/tests/ServerCancellationTests.scala

```diff
@@ -0,0 +1,55 @@
+package sttp.tapir.server.tests
+
+import cats.effect.IO
+import cats.effect.kernel.Async
+import cats.effect.syntax.all._
+import cats.syntax.all._
+import org.scalatest.matchers.should.Matchers._
+import sttp.client3._
+import sttp.monad.MonadError
+import sttp.tapir._
+import sttp.tapir.tests._
+
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.{Semaphore, TimeUnit}
+import scala.concurrent.duration._
+
+class ServerCancellationTests[F[_], OPTIONS, ROUTE](createServerTest: CreateServerTest[F, Any, OPTIONS, ROUTE])(implicit
+    m: MonadError[F],
+    async: Async[F]
+) {
+  import createServerTest._
+
+  def tests(): List[Test] = List({
+    val canceledSemaphore = new Semaphore(1)
+    val canceled: AtomicBoolean = new AtomicBoolean(false)
+    testServerLogic(
+      endpoint
+        .out(plainBody[String])
+        .serverLogic { _ =>
+          (m.eval(canceledSemaphore.acquire())) >> (async.sleep(15.seconds) >> pureResult("processing finished".asRight[Unit]))
```

Comment:
I think I don't understand the `async.sleep` - we're waiting on the semaphore, and releasing it only after the cancel signal comes in - do we need to still sleep after that?

Reply:
The semaphore will be acquired immediately here, the sleep is to simulate long processing. The semaphore is used to block the test code until backend's `onCancel` finishes setting the boolean.

Reply:
ah I mixed up the release/aquire directions ;) yeah it's fine of course :)